### PR TITLE
[IMP] core: use Python slots for records

### PR DESCRIPTION
**Pull request by rco-odoo**
Goal: restrict the attributes of a `BaseModel` instance to `env`, `_ids` and `_prefetch_ids`.
This way, one can only assign fields on a record; other assignments are programming errors.
This also reduces the memory footprint of records from 168 to 64 bytes (-62%).

### CODE REVIEW

File: addons/account/models/account_bank_statement.py [hunk 1]
```diff
@@ -1252,7 +1252,9 @@ def reconcile(self, lines_vals_list, to_check=False):
 
             # Update the payment date to match the current bank statement line's date.
             if counterpart_line.payment_id:
-                counterpart_line.payment_id.payment_date = self.date
+                # 'account.payment' object has no attribute 'payment_date'
+                # counterpart_line.payment_id.payment_date = self.date
+                pass
```
Comment (MiquelRForgeFlow) [inline addons/account/models/account_bank_statement.py:1257]:
```suggestion
                counterpart_line.payment_id.date = self.date
```

File: addons/point_of_sale/tests/test_anglo_saxon.py [hunk 2]
```diff
@@ -51,7 +51,8 @@ def setUp(self):
         self.cash_journal = self.env['account.journal'].create({'name': 'CASH journal', 'type': 'cash', 'code': 'CSH00'})
         self.sale_journal = self.env['account.journal'].create({'name': 'SALE journal', 'type': 'sale', 'code': 'INV00'})
         self.pos_config.invoice_journal_id = self.sale_journal
-        self.pos_config.journal_ids = [self.cash_journal.id]
+        # 'pos.config' object has no attribute 'journal_ids'
+        # self.pos_config.journal_ids = [self.cash_journal.id]
```
Comment (MiquelRForgeFlow) [inline addons/point_of_sale/tests/test_anglo_saxon.py:55]:
```suggestion
```
was correctly handled in https://github.com/odoo/odoo/commit/8becb17b8a3aa18b92629c4ddd33630ea4b1444e

File: addons/purchase_stock/tests/test_stockvaluation.py [hunk 3]
```diff
@@ -494,8 +494,9 @@ def test_valuation_from_increasing_tax(self):
         })
 
         self.product1.product_tmpl_id.categ_id.property_cost_method = 'fifo'
-        self.product1.product_tmpl_id.categ_id.valuation = 'real_time'
 
+        # 'product.category' object has no attribute 'valuation'
+        # self.product1.product_tmpl_id.categ_id.valuation = 'real_time'
```
Comment (MiquelRForgeFlow) [inline addons/purchase_stock/tests/test_stockvaluation.py:499]:
```suggestion
        self.product1.product_tmpl_id.categ_id.property_valuation = 'real_time'
```

File: addons/point_of_sale/tests/test_anglo_saxon.py [hunk 4]
```diff
@@ -51,7 +51,8 @@ def setUp(self):
         self.cash_journal = self.env['account.journal'].create({'name': 'CASH journal', 'type': 'cash', 'code': 'CSH00'})
         self.sale_journal = self.env['account.journal'].create({'name': 'SALE journal', 'type': 'sale', 'code': 'INV00'})
         self.pos_config.invoice_journal_id = self.sale_journal
-        self.pos_config.journal_ids = [self.cash_journal.id]
+        # 'pos.config' object has no attribute 'journal_ids'
```
Comment (pimodoo) [inline addons/point_of_sale/tests/test_anglo_saxon.py:54]: tu peux supprimer, car maintenant c'est "cash_payment_method", on a du oublier de retirer ca en ajoutant les payment_method

File: addons/stock_account/tests/test_stockvaluationlayer.py [hunk 5]
```diff
@@ -108,7 +108,9 @@ def _make_dropship_move(self, product, quantity, unit_cost=None):
             'picking_type_id': self.picking_type_out.id,
         })
         if unit_cost:
-            dropshipped.unit_cost = unit_cost
+            # 'stock.move' object has no attribute 'unit_cost'
+            # dropshipped.unit_cost = unit_cost
+            pass
```
Comment (MiquelRForgeFlow) [inline addons/stock_account/tests/test_stockvaluationlayer.py:113]:
```suggestion
            dropshipped.price_unit = unit_cost
```

File: addons/mrp/tests/test_workorder_operation.py [hunk 6]
```diff
@@ -1277,7 +1277,8 @@ def test_planning_6(self):
             'name': 'Manual Assembly',
             'time_cycle': 60,
         })
-        self.planning_bom.routing = mrp_routing_0
+        # 'mrp.bom' object has no attribute 'routing'
+        # self.planning_bom.routing = mrp_routing_0
```
Comment (MiquelRForgeFlow) [inline addons/mrp/tests/test_workorder_operation.py:1281]:
```suggestion
        self.planning_bom.routing_id = mrp_routing_0
```

File: addons/point_of_sale/tests/test_anglo_saxon.py (hunk 4 above)
Comment (rco-odoo) [reply addons/point_of_sale/tests/test_anglo_saxon.py:54]: Merci pour la review!

**Comment by rco-odoo** [comment]
@MiquelRForgeFlow thanks for your suggestions!  I will double-check them with my colleagues.

**Comment by sle-odoo** [comment]
@rco-odoo changes made here d29e569495149d0e3dd08a5152551958b2362f99

File: addons/hr_holidays/report/holidays_summary_report.py [hunk 7]
```diff
@@ -47,7 +47,7 @@ def _get_months(self, start_date):
             start_date += relativedelta(day=1, months=+1)
         return res
 
-    def _get_leaves_summary(self, start_date, empid, holiday_type):
+    def _get_leaves_summary(self, employee, start_date, holiday_type):
```
Comment (damdam-s) [inline addons/hr_holidays/report/holidays_summary_report.py:50]:
```suggestion
    def _get_leaves_summary(self, start_date, employee, holiday_type):
```
This avoid to change parameter order and do not break inheritance

Comment (rco-odoo) [reply addons/hr_holidays/report/holidays_summary_report.py:50]: You have got a point here.
I may keep the input parameters as before, but the output result *must change*.
So all the method overrides that *use* the output must be adapted.

File: addons/account/models/account_bank_statement.py (hunk 1 above)
Comment (qdp-odoo) [reply addons/account/models/account_bank_statement.py:1257]: indeed

File: addons/hr_holidays/report/holidays_summary_report.py [hunk 8]
```diff
@@ -76,29 +76,27 @@ def _get_leaves_summary(self, start_date, empid, holiday_type):
                     res[(date_from-start_date).days]['color'] = holiday.holiday_status_id.color_name
                 date_from += timedelta(1)
             count += holiday.number_of_days
-        self.sum = count
```
Comment (tivisse) [inline addons/hr_holidays/report/holidays_summary_report.py:79]: :+1:

Comment (rco-odoo) [reply addons/hr_holidays/report/holidays_summary_report.py:79]: Thanks for the feedback!

**Comment by rco-odoo** [comment]
@robodoo r+

**Pull request closed without merging**